### PR TITLE
Setup Product Construction Service

### PR DESCRIPTION
**Pull request by dkurepa**
https://github.com/dotnet/arcade-services/issues/3162
https://github.com/dotnet/arcade-services/issues/3158
https://github.com/dotnet/arcade-services/issues/3173

This PR includes:
 - a new aspire project that currently just has an empty API.
 - a bicep template for provisioning the whole staging environment
 - yaml for a deployment pipeline using the blue/green deployment approach
 - sets up Application Insights logging

### Release Note Category
- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements

### Release Note Description
Setup the Product Construction Service

### CODE REVIEW

File: eng/deployment/product-construction-service-deploy.ps1
Comment (premun) [inline eng/deployment/product-construction-service-deploy.ps1:1]: Please put a comment on top of the file explaining what this file does in some high level way

File: docs/ProductConstructionServiceDevGuide.md
Comment (premun) [inline docs/ProductConstructionServiceDevGuide.md:1]: Can you merge this with `instructions.md` and call it `Readme.md` and just put it in the PCS folder?

File: src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs:12]: Should we just use the `/health` one provided by ASP.NET?

File: src/ProductConstructionService/ProductConstructionService.Api/Dockerfile
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Api/Dockerfile:21]: I think it's better to use relative paths so that we only need to change `WORKDIR /src` on top and move everything

```suggestion
COPY ["/src/ProductConstructionService/ProductConstructionService.ServiceDefaults/ProductConstructionService.ServiceDefaults.csproj", "./ProductConstructionService/ProductConstructionService.ServiceDefaults/"]
COPY ["src/ProductConstructionService/ProductConstructionService.Api/ProductConstructionService.Api.csproj", "./ProductConstructionService/ProductConstructionService.Api/"]
RUN dotnet restore "/src/ProductConstructionService/ProductConstructionService.Api/ProductConstructionService.Api.csproj"
COPY ["src/ProductConstructionService", "."]
RUN dotnet build "./ProductConstructionService.Api/ProductConstructionService.Api.csproj" -c $BUILD_CONFIGURATION -o /app/build --no-restore
```

File: azure-pipelines.yml
Comment (premun) [inline azure-pipelines.yml:164]:
```suggestion
        - powershell: |
            .\eng\common\build.ps1 -restore
            .\.dotnet\dotnet install workload aspire
          displayName: Install .NET and Aspire
```

File: Directory.Packages.props
Comment (dkurepa) [reply Directory.Packages.props:120]: there are no stable versions, aspire is still in preview technically

Comment (premun) [reply Directory.Packages.props:120]: But this is about `OpenTelemetry`:
https://www.nuget.org/packages/OpenTelemetry.Instrumentation.AspNetCore

File: azure-pipelines-product-construction-service.yml
Comment (dkurepa) [reply azure-pipelines-product-construction-service.yml:2]: Added the non production trigger. The PR triggers are configured in the repo settings, and it's currently off

File: eng/deployment/product-construction-service-deploy.ps1
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1:4]: added

File: docs/ProductConstructionServiceDevGuide.md
Comment (dkurepa) [reply docs/ProductConstructionServiceDevGuide.md]: done

File: src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs:12]: I think we'll want to have a specialized one for the service, that we can configure to say `ready` after initializing the VMR, for example

Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs:12]: Let's add it then. It might happen quite far in the future or not at all and until then, I think we might be missing out on whatever the health endpoint does - I imagine it might be smart but don't know - would be nice to find out maybe

File: src/ProductConstructionService/Readme.md
Comment (premun) [inline src/ProductConstructionService/Readme.md:8]: Should it be this?
```suggestion
 - `az group create --location northcentralusstage --name product-construction-service`
```

File: src/ProductConstructionService/Readme.md
Comment (premun) [inline src/ProductConstructionService/Readme.md:5]: I am wondering what's the point of having these commands in a markdown rather than just have `Create-Azure-Resources.ps1`?

File: src/ProductConstructionService/Readme.md
Comment (premun) [reply src/ProductConstructionService/Readme.md:8]: Also please put a comment why this location

File: src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs
Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs:12]: Also - can there be more? Maybe we keep the default and just add a second one later?

File: azure-pipelines-product-construction-service.yml
Comment (premun) [reply azure-pipelines-product-construction-service.yml:2]: It is still better to add `pr: none`. For instance, this will work for internal PRs in AzDO if we ever make some.

It's also explicit when reading the YAML - that we don't expect this to run on PRs. So you don't need to go in AzDO and search for the pipelines etc

File: src/ProductConstructionService/Readme.md
Comment (dkurepa) [reply src/ProductConstructionService/Readme.md:8]: I think that step might not be doable with azure cli at all, as it has input validation that does not include that location.
Also creating a RG on the portal doesn't allow you to create a resource group with that location. Do we need a special feature flag in our subscription?

Comment (premun) [reply src/ProductConstructionService/Readme.md:8]: I sent a question to the manager there asking if we can use this in other locations already

Comment (premun) [reply src/ProductConstructionService/Readme.md:8]: My experimental app has this YAML description:
```yaml
id: /subscriptions/0b38483d-f273-4816-832b-59fea3a17a23/resourceGroups/backflow-service/providers/Microsoft.App/containerapps/backflow-service-app
identity:
  type: None
location: North Central US (Stage)
name: backflow-service-app
```

Comment (dkurepa) [reply src/ProductConstructionService/Readme.md:8]: we probably won't be able to create the resource group with azure CLI, I'll update the instructions. Hopefully bicep doesn't give us trouble with that.

Comment (premun) [reply src/ProductConstructionService/Readme.md:8]: I am not sure what is the right string for that location, yeah

**Comment by premun** [comment]
@dkurepa do you think it would be worth the hassle to try to at least build the docker image? I wonder how often we'll be making mistakes there. Since this is not run in local and PR, we will only find out after merging (maybe that's okay too?).

**Comment by dkurepa** [comment]
> @dkurepa do you think it would be worth the hassle to try to at least build the docker image in PRs? I wonder how often we'll be making mistakes there. Since this is not run in local and PR, we will only find out after merging (maybe that's okay too?).

I think that's a valid concern, I see that as a part of PR validation

**Comment by premun** [comment]
> > @dkurepa do you think it would be worth the hassle to try to at least build the docker image in PRs? I wonder how often we'll be making mistakes there. Since this is not run in local and PR, we will only find out after merging (maybe that's okay too?).
> 
> I think that's a valid concern, I see that as a part of PR validation

Let's not block on this and log an issue and we will see how much of a problem this is

**Comment by dkurepa** [comment]
> > > @dkurepa do you think it would be worth the hassle to try to at least build the docker image in PRs? I wonder how often we'll be making mistakes there. Since this is not run in local and PR, we will only find out after merging (maybe that's okay too?).
> > 
> > 
> > I think that's a valid concern, I see that as a part of PR validation
> 
> Let's not block on this and log an issue and we will see how much of a problem this is

I added the condition to the new pipeline, we can just run it as a part of PR validation?

File: src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Api/Controllers/ServiceStatusController.cs:12]: Yeah, that might be the best for now, I'll delete this controller, and set the Health probe to use the `/health` endpoint.

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:7]: I don't think we want this, we want to include `main` only.
People pushing in their dev branches should not trigger build automatically. I'd say.

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:53]: I'd rather this said `if branch is main`

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:48]:
```suggestion
      displayName: Set new docker image tag variable
```

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:51]:
```suggestion
      displayName: Build docker image
```

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:37]: We build the rest of the repo on Windows only, we use Windows locally too. Why Ubuntu?

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:11]:
```suggestion
  value: product-construction-service
```

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:13]:
```suggestion
  value: product-construction-service.api
```

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:67]:
```suggestion
            scriptType: pscore
            scriptLocation: scriptPath
            scriptPath: $(Build.SourcesDirectory)/eng/deployment/product-construction-service-deploy.ps1
            arguments: -resourceGroupName $(resourceGroupName) -containerappName $(containerappName) -newImageTag $(newDockerImageTag) -containerRegistryName $(containerRegistryName) -imageName $(containerName) -subscriptionName $(subscriptionName) -containerappEnvironmentName "acaexjpe3hisjmv2e"
          displayName: Product Construction Service Deploy
```

File: eng/deployment/product-construction-service-deploy.ps1
Comment (premun) [inline eng/deployment/product-construction-service-deploy.ps1:1]:
```suggestion
# This script deploys the Product Construction Service using the blue/green deployment pattern.
```

File: eng/deployment/product-construction-service-deploy.ps1
Comment (premun) [inline eng/deployment/product-construction-service-deploy.ps1:2]:
```suggestion
# The script determines the color of the currently active revision, deactivates the old inactive revision, 
```

File: eng/deployment/product-construction-service-deploy.ps1
Comment (premun) [inline eng/deployment/product-construction-service-deploy.ps1:50]: What happens if the image is not there? Will this be a visible message?

Comment (premun) [reply eng/deployment/product-construction-service-deploy.ps1:50]: Or why the `| Out-Null`?

Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1:50]: We`ll get an error then, the `Out-Null` only affects stdout. I added it because the output of these az cli commands is a full json view of the affected resource, which is quite long, and convolutes the output

File: src/ProductConstructionService/Readme.md
Comment (dkurepa) [reply src/ProductConstructionService/Readme.md]: I modified the provisioning script as we talked about. We now create the Container Environment and the app in the North Central US Stage location, while the rest of the resources is in the normal, North Central US location

File: src/ProductConstructionService/Readme.md
Comment (dkurepa) [reply src/ProductConstructionService/Readme.md:5]: Added a script `provision.ps1` and updated the instructions

File: Directory.Packages.props
Comment (premun) [inline Directory.Packages.props:119]: There is 1.7.0 of these as well

Comment (dkurepa) [reply Directory.Packages.props]: fixed

File: azure-pipelines-product-construction-service.yml
Comment (dkurepa) [reply azure-pipelines-product-construction-service.yml:7]: yeah, you're right. If someone wants to trigger the pipeline from their dev branch, they should do so explicitly

File: azure-pipelines-product-construction-service.yml
Comment (dkurepa) [reply azure-pipelines-product-construction-service.yml:49]: if we do that, we won't be able to push to the staging environment from dev branches. I think this is inline with what we have for the main pipeline

File: azure-pipelines-product-construction-service.yml
Comment (dkurepa) [reply azure-pipelines-product-construction-service.yml]: fixed everywhere

File: azure-pipelines-product-construction-service.yml
Comment (dkurepa) [reply azure-pipelines-product-construction-service.yml]: fixed everywhere

File: azure-pipelines-product-construction-service.yml
Comment (dkurepa) [reply azure-pipelines-product-construction-service.yml]: I wasn't able to change the Docker daemon to Linux on the `windows-latest` image I originally tried using. I'll try some other image, it might be configured to Linux by default

Comment (premun) [reply azure-pipelines-product-construction-service.yml]: Ah, nevermind then

File: azure-pipelines-product-construction-service.yml
Comment (dkurepa) [reply azure-pipelines-product-construction-service.yml]: fixed

File: azure-pipelines-product-construction-service.yml
Comment (premun) [inline azure-pipelines-product-construction-service.yml:30]: FYI, you can also have a single-stage YAML, so there's not `stages:` but just
`jobs:` directly at the top level

File: src/ProductConstructionService/ProductConstructionService.Api/ProductConstructionService.Api.csproj
Comment (premun) [inline src/ProductConstructionService/ProductConstructionService.Api/ProductConstructionService.Api.csproj:24]: Is this still needed?

**Review by premun — DISMISSED** [review]

File: src/ProductConstructionService/ProductConstructionService.Api/ProductConstructionService.Api.csproj
Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Api/ProductConstructionService.Api.csproj]: Removed

**Review by premun — APPROVED** [review]

**Comment by premun** [comment]
@dkurepa we should add the new pipeline to the build monitor so that we get notified when it goes red

**Review by premun — APPROVED** [review]

**Pull request merged**
